custom_test.go

```diff
@@ -120,66 +103,123 @@ func (ia *IntSlice) Unpack(r io.Reader, length int, opt *Options) error {
 			}
 			return err
 		}
-		*ia = append(*ia, value)
 		if value == 0 {
 			break
 		}
+		*ia = append(*ia, value)
 	}
 	return nil
 }
 
-func (ia *IntSlice) Size(opt *Options) int {
+func (ia *SliceUInt8) Size(opt *Options) int {
 	return len(*ia) + 1
 }
 
-func (ia *IntSlice) String() string {
+func (ia *SliceUInt8) String() string {
 	panic("not implemented")
 }
 
-func TestCustomLength(t *testing.T) {
-	var buf bytes.Buffer
-	i := make(IntSlice, 0)
-	i = append(i, 128)
-	i = append(i, 64)
-	i = append(i, 32)
-
-	if err := Pack(&buf, &i); err != nil {
-		t.Fatal(err)
-	}
-	if !bytes.Equal(buf.Bytes(), []byte{128, 64, 32, 0}) {
-		t.Fatal("error packing custom int array")
-	}
-	var i2 IntSlice
-	if err := Unpack(&buf, &i2); err != nil {
-		t.Fatal(err)
-	}
-	if i2[0] != 128 {
-		t.Fatal("error unpacking custom int array")
-	}
-}
-
-type IntSliceStruct struct {
-	I IntSlice
+type SliceUInt8Struct struct {
+	I SliceUInt8
 	N uint8 // A field after to ensure the length is correct.
 }
 
-func TestCustomLengthStruct(t *testing.T) {
-	var buf bytes.Buffer
-	i := IntSliceStruct{
-		I: IntSlice{128, 64, 32},
-		N: 192,
-	}
-	if err := Pack(&buf, &i); err != nil {
-		t.Fatal(err)
-	}
-	if !bytes.Equal(buf.Bytes(), []byte{128, 64, 32, 0, 192}) {
-		t.Fatal("error packing custom int array struct")
-	}
-	var i2 IntSliceStruct
-	if err := Unpack(&buf, &i2); err != nil {
-		t.Fatal(err)
-	}
-	if i2.I[0] != 128 || i2.N != 192 {
-		t.Fatal("error unpacking custom int array struct")
+func TestCustomTypes(t *testing.T) {
+	testCases := []struct {
+		name        string
+		packObj     interface{}
+		emptyObj    interface{}
+		expectBytes []byte
+		expectPanic bool // True for unimplemented features
+	}{
+		{
+			name:        "CustomType",
+			packObj:     newInt3(3),
+			emptyObj:    newInt3(0),
+			expectBytes: []byte{0, 0, 3},
+		},
+		{
+			name:        "CustomType-Big",
+			packObj:     newInt3(4000),
+			emptyObj:    newInt3(0),
+			expectBytes: []byte{0, 15, 160},
+		},
+		{
+			name:        "CustomTypeStruct",
+			packObj:     &Int3Struct{3},
+			emptyObj:    &Int3Struct{},
+			expectBytes: []byte{0, 0, 3},
+		},
+		{
+			// Test is wrong, but expectFail() is not available:
+			// https://github.com/golang/go/issues/25951
+			//
+			name:        "ArrayOfCustomType",
+			packObj:     &[2]Int3{3, 4},
+			emptyObj:    &[2]Int3{},
+			expectBytes: []byte{0, 0, 0, 3, 0, 0, 0, 4}, // FIXME: INCORRECT Should panic.
```

Comment:
Calling out this test. It should panic, because the feature is not implemented. IMO it is probably out of scope for this PR, because it is an existing missing feature. I'm not sure where it is best to detect and throw the panic. Suggestions? Maybe in `struct.go/prep()`? https://github.com/lunixbochs/struc/blob/02e4c2afbb2ac4bae6876f52c8273fc4cf5a4b0a/struc.go#L31

Reply:
Fixed and now handled in `struct.go/prep()`